datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -1307,6 +1309,39 @@ impl<'a, S: SimplifyInfo> TreeNodeRewriter for Simplifier<'a, S> {
                 ExprSimplifyResult::Simplified(expr) => Transformed::yes(expr),
             },
 
+            Expr::AggregateFunction(AggregateFunction {
```

Comment:
It seems like in order to be able to make this API avoid copying, it would need to pass the `distinct`, `filter`, etc fields in and then get them back

Another thing we could do is to use `Transformed` directly

What if we made something like

```rust
struct AggregateFunctionsArgs {
  args: Vec<Expr>,
  distinct: bool,
  filter: Option<Expr>,
  order_by: ...
}
```

And then the call to simplify to be 

```rust
trait AggreagateUDFImpl { 
...
    fn simplify(
        &self,
        agg_args: AggregateFunctionsArgs,
    ) -> Result<Transformed<AggregateFunctionsArgs> {
        Ok(Transformed::no(agg_args))
    }
```

Reply:
I agree with `AggregateFunctionsArgs`, consistency part got me on board 😀

I'm not sure about `Transformed` part, reason being is that `ExprSimplifyResult` returns two types, `Expr` and currently `Vec<Expr>`

in that case `simplify` signature should be changed to: 

```rust
trait AggreagateUDFImpl { 
...
    fn simplify(
        &self,
        agg_args: AggregateFunctionsArgs,
    ) -> Result<Transformed<Expr> {
        let original : Expr = agg_args.into();
        Ok(Transformed::no(original))
    }
```

Thus simplify function should return 're-composed' original function or new function ...

Anyway, I'll have a look what can be done, we can discuss